Alert: company info incomplete

weedySeaDragon opened the pull request:
### PT Story: Email members with incomplete company information, ask them to complete it
https://www.pivotaltracker.com/story/show/151655795

### Changes proposed in this pull request:
1. Created the `CompanyInfoIncompleteAlert` class
  -  with a _draft_ schedule of `[2, 9, 14, 30, 60] days after the first person became a member`
   - locale entries
   -  show the items missing in the email

2. added an abstract class `AbstractOppositeRequirements`  as a parent class for those simple classes that are the _opposite_ of other requirements since there are now 2 classes that are opposites (`RequirementsForHBrandingFeeNotDue` and `RequirementsForCoInfoNotComplete`).  Pull out common behavior.

3. added method `complete?` to `Company` that calls 
```
RequirementsForCoInfoComplete.requirements_met? company: self
``` 
   to check if it's complete 

4. Reorganized methods in Company so that scopes are at the top, then class methods, then instance methods, per RuboCop standards.

### Screenshots (Optional):
**[sv]**
<img width="631" alt="email-members-co-info-incomplete-sv-3" src="https://user-images.githubusercontent.com/673794/52891161-fe41c500-313d-11e9-9b6c-3eaffc4e1721.png">


---
**[en]**
<img width="628" alt="email-members-co-info-incomplete-en-3" src="https://user-images.githubusercontent.com/673794/52891163-06016980-313e-11e9-92e5-4a79f5fd8b36.png">




### Ready for review:
@thesuss 
@patmbolger

weedySeaDragon committed:
reorganize: class methods at top per RuboCop standards; add whitespace

weedySeaDragon committed:
DRAFT locale entries

weedySeaDragon committed:
parent class for Requirements that are just the opposite of another requirement

weedySeaDragon committed:
refactor so it's a subclass of the AbstractOppositeRequirements class

weedySeaDragon committed:
add step that sets the company name to a blank

weedySeaDragon committed:
add Requirements for Company Info Complete (or not complete) classes

weedySeaDragon committed:
add method complete? (uses Requirements class); branding_start_date

weedySeaDragon committed:
update h-branding-fee-not-due spec to check for opposite class

weedySeaDragon committed:
specs for RequirementsForCoInfo[Not]Complete classes

weedySeaDragon committed:
add email view and mailer method + spec and preview

weedySeaDragon committed:
remove branding_start_date; no needed for Alert

weedySeaDragon committed:
feature for Company info incomplete alert

weedySeaDragon committed:
Company info not complete alert and spec

weedySeaDragon committed:
add company_info_incomplete to Mailer preview

weedySeaDragon committed:
correct the comment for CompanyInfoIncompleteAlert

weedySeaDragon committed:
add to load_conditions; make schedule arrays std

thesuss commented on config/locales/sv.yml:
```diff
@@ -1054,7 +1057,15 @@ sv:
+          company_link_login_msg: &co_link_login " (du kan behöva logga in först innan du ser knappen för att betala)."
+
+      co_info_incomplete:
+        subject: Please complete the company information
```
Var god färdigställ din företagssida

thesuss commented on config/locales/sv.yml:
```diff
@@ -1054,7 +1057,15 @@ sv:
+        subject: Please complete the company information
+        greeting: *email_greeting
+        message_text:
+          complete_the_info: "You need to complete the info for %{co_name}. If you don't it won't show up on the map or in google searches"
```
Q: What happens if name is empty? 

Du behöver färdigställa informationen om %{co_name}. Om du inte gör det syns du inte i sökningar. Saknar du adress syns du heller inte på kartan.

weedySeaDragon replied:
If the company name is blank, then it shows` company_needs_name: "Org nummer %{co_number}"`

where `%{co_name}` is.  (That logic is in the mailer view app/views/member_mailer/company_info_incomplete.html.haml)

So it would say:

Du behöver färdigställa informationen om Org nummer %{co_number}. Om du inte gör det syns du inte i sökningar. Saknar du adress syns du heller inte på kartan.

thesuss replied:
Ah I see that now. Ok. Will update that row

thesuss commented on config/locales/sv.yml:
```diff
@@ -1054,7 +1057,15 @@ sv:
+        message_text:
+          complete_the_info: "You need to complete the info for %{co_name}. If you don't it won't show up on the map or in google searches"
+          company_needs_name: "Org nummer %{co_number}"
+          company_link_login_msg: "Here's the link to the company so you can edit it: "
```
Här är länken där du kan redigera ditt företag:

weedySeaDragon committed:
locale entry updates per Susanna

thesuss commented on config/locales/sv.yml:
```diff
@@ -1054,7 +1057,15 @@ sv:
+        greeting: *email_greeting
+        message_text:
+          complete_the_info: "You need to complete the info for %{co_name}. If you don't it won't show up on the map or in google searches"
+          company_needs_name: "Org nummer %{co_number}"
```
"företag: %{co_number}"

weedySeaDragon committed:
add method missing_region?

weedySeaDragon committed:
use Company method missing_region? so implementation is hidden

weedySeaDragon committed:
partial for 'you may need to login first' with link to login page

weedySeaDragon committed:
shared spec for 'you may need to login first'; will be used in mult. emails

weedySeaDragon committed:
list what is missing in the email; spec, draft locale entries

weedySeaDragon committed:
update preview with example missing name and region

weedySeaDragon committed:
sv locale entry for company_needs_name -> foretag

weedySeaDragon commented:
@thesuss I updated the locale with your info.  Also I *added* 4 locale entries to be able to show what is missing.
1.  I updated the screen shot in the PR description (above)
2. I added these 4 entries:
```
          what_is_missing: "Här är vad som saknas:"
          co_name_missing: "det finns inget namn för företaget"
          co_region_missing: "regionen för företaget tillhandahålls inte"
          please_complete: "Vänligen lägg till ovanstående information så att företagsinformationen är klar."
```

Just FYI in case you want to look at them in the locale file, they start on line 1068:
https://github.com/AgileVentures/shf-project/blob/c87a65e4a5a6bbc7ef924d3b73be015bd241247e/config/locales/sv.yml#L1068

thesuss commented on config/locales/sv.yml:
```diff
@@ -1054,7 +1057,19 @@ sv:
+          complete_the_info: "Du behöver färdigställa informationen om %{co_identifier}. Om du inte gör det syns du inte i sökningar. Saknar du adress syns du heller inte på kartan."
+          company_needs_name: "företag: %{co_number}"
+          what_is_missing: "Här är vad som saknas:"
+          co_name_missing: "det finns inget namn för företaget"
```
Företagsnamn

thesuss commented on config/locales/sv.yml:
```diff
@@ -1054,7 +1057,19 @@ sv:
+          company_needs_name: "företag: %{co_number}"
+          what_is_missing: "Här är vad som saknas:"
+          co_name_missing: "det finns inget namn för företaget"
+          co_region_missing: "regionen för företaget tillhandahålls inte"
```
Postadress

weedySeaDragon replied:
We are specifically checking that the **region** is not nil.  They might have other parts of the address in there, but if they left out the Region, we call it "incomplete".

So - I'm just checking that the Swedish wording reflects that it's the _region_.

Is "postadress" really what you want there?

thesuss replied:
Yes, they dont need to know what we are checking :) 
and I'm fine with them not being displayed on the map if they chose not to be. But I'm not fine with them not having a postadress :D

thesuss commented on config/locales/en.yml:
```diff
@@ -1061,6 +1064,19 @@ en:
+          complete_the_info: "You need to complete the info for %{co_identifier}. If you don't it won't show up on the map or in Google searches"
+          company_needs_name: "company number %{co_number}"
+          what_is_missing: "Here is what's missing:"
+          co_name_missing: "there is no name for the company"
```
Company name

thesuss commented on config/locales/en.yml:
```diff
@@ -1061,6 +1064,19 @@ en:
+          company_needs_name: "company number %{co_number}"
+          what_is_missing: "Here is what's missing:"
+          co_name_missing: "there is no name for the company"
+          co_region_missing: "the region for the company is not provided"
```
Address

patmbolger commented on app/models/reqs_updaters/requirements_for_co_info_complete.rb:
```diff
@@ -0,0 +1,36 @@
+  def self.requirements_met?(args)
+    company = args[:company]
+
+    !( company.name.blank? || company.missing_region? )
```
Would be useful to have a comment here that states that if the requirements for company "complete" status change, then not only this logic but the logic for `Company.complete` also has to change (need a comment in that method as well, pointing here).

patmbolger commented on app/models/conditions_response/company_info_incomplete_alert.rb:
```diff
@@ -0,0 +1,40 @@
+
+      earliest_date = company.earliest_current_member_fee_paid
+
+      day_to_check = self.class.days_today_is_away_from(earliest_date, timing)
```
Why `self.class`?  Why not explicitly reference the class?

weedySeaDragon replied:
makes it more flexible.  That way it can easily be changed/subclassed/etc.  Is a general coding practice. No need to be so specific, no need to restrict the code (= no need to put in the class name) if you don't have to for some good reason. 
#SOLID  "O" - open to change

weedySeaDragon replied:
(@patmbolger - just a reminder to give some feedback when you can. didn't want this to get lost in all the other stuff)

patmbolger commented on app/models/reqs_updaters/abstract_opposite_requirements.rb:
```diff
@@ -0,0 +1,49 @@
+
+
+  # This requires the same arguments as RequirementsForHBrandingFeeDue since
+  # it sends those arguments to RequirementsForHBrandingFeeDue.requirements_met?
```
comment needs updating

patmbolger commented on app/models/reqs_updaters/abstract_opposite_requirements.rb:
```diff
@@ -0,0 +1,49 @@
+  end
+
+
+  # This is always just the opposite of if the required info for a company is incomplete
```
comment needs updating

patmbolger commented on app/models/reqs_updaters/requirements_for_co_info_complete.rb:
```diff
@@ -0,0 +1,36 @@
+#       This is a very simple class because the requirements are currently very simple.
+#       The importance is that
+#        it is the only place that code needs to be touched if the rules for
+#        when the all the required information for a company is complete
```
No - we also need to touch the Company `complete` scope.

patmbolger reviewed (commented):
Please see comments

weedySeaDragon committed:
locale updates: co_name_missing, co_region_missing

weedySeaDragon committed:
comments: if def. of 'complete' changes, changes in 2 places are required

weedySeaDragon committed:
fix comments in AbstractOppositeRequirements class

weedySeaDragon committed:
Merge branch 'develop' of https://github.com/AgileVentures/shf-project into sprint61-alert-company-incomplete-#151655795

patmbolger reviewed (approved):
Looks good!
